Fix save/load states in Emscripten build

jbaicoianu opened the pull request:
Slightly rearranges the ```running_machine::run()``` function so that the Emscripten build can share the same codepath, allowing it to access the scheduler to process any save/load commands that may have been queued.

cuavas commented on src/emu/machine.cpp:
```diff
@@ -424,6 +410,31 @@ int running_machine::run(bool quiet)
+	g_profiler.start(PROFILER_EXTRA);
+
+	// execute CPUs if not paused
+	if (!m_paused) {
```
This file uses Allman style, please be consistent with it.

cuavas commented on src/emu/machine.cpp:
```diff
@@ -424,6 +410,31 @@ int running_machine::run(bool quiet)
+
+	// execute CPUs if not paused
+	if (!m_paused) {
+		attotime stoptime(m_scheduler.time() + frametime);
```
If this can't change it should be const.

cuavas commented on src/emu/machine.cpp:
```diff
@@ -424,6 +410,31 @@ int running_machine::run(bool quiet)
+	// execute CPUs if not paused
+	if (!m_paused) {
+		attotime stoptime(m_scheduler.time() + frametime);
+		while (m_scheduler.time() < stoptime) {
```
This loop will never be entered in the non-Emscripten case because `m_scheduler.time() < stoptime` will never be true when `stoptime` is `m_scheduler.time() + attotime::zero`.  Also, if you want the loop to be entered at least once, should use `do ... while ()`.

cuavas commented on src/emu/machine.h:
```diff
@@ -215,6 +215,7 @@ class running_machine
 
 	// immediate operations
 	int run(bool quiet);
+	int run_frame(attotime frametime=attotime::zero);
```
Should this really be here, not protected or something?  Also, default argument values are a source of problems.  The name is also confusing, as it runs one or more scheduler timeslices, not really a "frame" by any definition.

galibert commented:
In addition I have a fundamental problem with that patch: it tries to
enshrine the "run for one frame" concept which we're trying to get away
from.  An emulated system can have multiple screens with different
framerates.  Also, we don't always know at the start of the frame when it's
going to end (zx80/81 where sync is software-controlled, atari st where
frequency changes are done often for overscan, arcade games that can change
resolution, etc).  So committing to a frame duration is really
problematic...

  OG.


On Thu, Jul 13, 2017 at 12:09 PM, James Baicoianu <notifications@github.com>
wrote:

> Slightly rearranges the running_machine::run() function so that the
> Emscripten build can share the same codepath, allowing it to access the
> scheduler to process any save/load commands that may have been queued.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/mamedev/mame/pull/2470
> Commit Summary
>
>    - Fix save/load states in Emscripten build
>
> File Changes
>
>    - *M* src/emu/machine.cpp
>    <https://github.com/mamedev/mame/pull/2470/files#diff-0> (61)
>    - *M* src/emu/machine.h
>    <https://github.com/mamedev/mame/pull/2470/files#diff-1> (1)
>
> Patch Links:
>
>    - https://github.com/mamedev/mame/pull/2470.patch
>    - https://github.com/mamedev/mame/pull/2470.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mamedev/mame/pull/2470>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AI0i8eOvrsm8m5awU2ZXKuTkJEnic7o_ks5sNezOgaJpZM4OWx_E>
> .
>

DavidHaywood commented:
yeah, I have to agree with OG on this one.

the issues he mentions are already real ones (racedrivpan side screens don't run at the same refresh as the main monitor and we have an ugly hack just to stop the updates from failing as it is)

so code tied the the machine refresh / frames is going to be problematic.

jbaicoianu commented:
@cuavas thanks for the feedback, I've incorporated most of your suggestions.  I kept the function public though - since this is run from the Emscripten main loop which is executed outside of the class scope, we need to be able to call this function from outside, as seen here: https://github.com/jbaicoianu/mame/blob/emscripten-fix-loadsave/src/emu/machine.cpp#L1337-L1340

With respect to the feedback from @galibert and @DavidHaywood - I'm open to feedback about how to best accomplish this given the implementation details of the target platform (Emscripten).  The fundamental difference between how an Emscripten app works compared to a native app is that while in the native app it's perfectly normal to have a tight loop executing infinitely on the main thread, in the browser environment this is a huge problem, because then your application code locks the browser's main thread and it can't even perform basic draw functions for that page.  Emscripten works around this by registering a custom main function which is called at regular intervals, and the most common / best-performing option here is to sync that with frame updates.

Currently to my knowledge the web builds don't support multiple screen outputs anyway.  This pull request isn't really introducing any new code or new limitations, it's just rearranging what's already there to allow more functionality to be accessed by the web builds.

If you have some suggestions for how to better approach this problem, I would be glad to incorporate them.

jbaicoianu commented:
Thought about it a bit more and I found a way to do it with less changes, and keeps the run-for-one-frame code isolated to the Emscripten-specific code.  So now there should be no significant changes for the native app, aside from the addition of a new public function ```running_machine::run_timeslice()``` which contains all the code which used to be inside the main app loop.

cuavas commented:
Is it actually necessary for the Emscripten event loop to be outside class scope?

cuavas commented on src/emu/machine.cpp:
```diff
@@ -1322,16 +1329,13 @@ static running_machine * jsmess_machine;
 
 void js_main_loop()
```
Is it necessary for this to be a free function?  It seems like more of a case for a static member function of the `running_machine`.

ajrhacker commented:
The new version looks much improved. Once the namespace issue and one K&R-style brace left in js_main_loop are fixed, this should be good to go.

jbaicoianu commented:
Ok, I've fixed the indentation and moved all of the standalone global functions involved in the Emscripten port to be static member functions of the ```running_machine``` class.  I also had to change the symbol names of the exported functions, so that these functions remain callable from JS scripts.

Now that I've made these changes though, the load functionality seems to have stopped working.  Please let me know if the new changes look like what you had in mind for the code restructure, meanwhile I'll track down why that functionality no longer works.

cuavas commented:
Yep, that looks much better.  If you can just get it working while looking more-or-less like that we should be good to go.

jbaicoianu commented:
Ok, I think that should do it.  As it turns out, there were some problems with subtly inverted logic introduced when I split the timeslice code out into a separate ```running_machine::run_timeslice``` function.  

Now that everything is a static member function attached to ```running_machine``` that separation is no longer necessary, so I've put the native codepath back the way it was before I started, and now everything is self-contained in the ```running_machine::emscripten_main_loop``` function.

As an added bonus, this now allows for proper shutdown of the running machine when running in the Emscripten environment - previously, attempts to exit the program were just being ignored.

cuavas commented:
Thanks, that's quite a bit cleaner.
